src/rgw/rgw_qos.h

```diff
@@ -0,0 +1,256 @@
+      auto& it = find_or_create(key);
+      auto curr_ts = curr_timestamp.time_since_epoch();
+      std::unique_lock lock(it.ts_lock);
+      increase_tokens(it, curr_ts, qos_info);
+      if(is_read)
+        return should_rate_limit_read(it, qos_info->max_read_ops, qos_info->max_read_bytes);
+      return should_rate_limit_write(it, qos_info->max_write_ops, qos_info->max_write_bytes);
+    }
+    void giveback_tokens(const char *method, const std::string& key)
+    {
+      bool is_read = is_read_op(method);
+      auto& it = find_or_create(key);
+      std::unique_lock lock(it.ts_lock);
+      if(is_read) {
+        it.read.ops += fixed_float_rgw_ratelimit;
+      } else {
+        it.write.ops += fixed_float_rgw_ratelimit;
+      }
+      
+    }
+    void increase_bw(const char *method,const std::string& key, int64_t amount, const RGWQoSInfo* info, const DoutPrefixProvider* dpp) {
+      if(key.empty() || key.length() == 1 || !info->enabled)
+        return;
+      bool is_read = is_read_op(method);
+      if ((is_read && !info->max_read_bytes) || (!is_read && !info->max_write_bytes))
+        return;
+      auto& it = find_or_create(key);
+      std::unique_lock lock(it.ts_lock);
+      // we don't want the tenant to be with higher debt than 120 seconds(2 min) of its limit
+      if (is_read)
+        it.read.bw = std::max(it.read.bw - amount * fixed_float_rgw_ratelimit,info->max_read_bytes * fixed_float_rgw_ratelimit * -2);
+      else {
+        it.write.bw = std::max(it.write.bw - amount * fixed_float_rgw_ratelimit,info->max_write_bytes * fixed_float_rgw_ratelimit * -2);
+      }
+      ldpp_dout(dpp, 20) << "rate limiting read bytes is: " << it.read.bw <<
+                  " write bytes is: " <<  it.write.bw << dendl;
+      ldpp_dout(dpp, 20) << "rate limiting key is: " << key << " bytes added: " << amount <<
+      " method is: " << method << dendl;
+    }
+    void clear() {
+      qos_entries.clear();
+    }
+};
+// This class purpose is to hold 2 RateLimiter instances, one active and one passive.
+// once the active has reached the watermark for clearing it will call the replace_active() thread using cv
+// The replace_active will clear the previous RateLimiter after all requests to it has been done (use_count() > 1)
+// In the meanwhile new requests will come into the newer active
+class ActiveRateLimiter : public DoutPrefix  {
+  std::atomic_uint8_t stopped = {false};
+  std::condition_variable cv;
+  std::mutex cv_m;
+  std::thread runner;
+  std::atomic_bool replacing = false;
+  std::atomic_uint8_t current_active = 0;
+  std::shared_ptr<RateLimiter> qos[2];
+  void replace_active() {
+    using namespace std::chrono_literals;
+    std::unique_lock<std::mutex> lk(cv_m);
+    while(!stopped) {
+      cv.wait(lk);
+      current_active = current_active ^ 1;
+      ldpp_dout(this, 20) << "replacing active qos data structure" << dendl;
+      while (!stopped && qos[(current_active ^ 1)].use_count() > 1 ) {
```

Comment:
what if `use_count()` never goes to 1? as requests keep on arriving

Reply:
Because of this line:       current_active = current_active ^ 1;
New requests will come to the other map as it is swapping the active map